packages/client/hmi-client/src/components/responsive-matrix/matrix.vue

```diff
@@ -0,0 +1,902 @@
+				return [];
+			}
+		},
+		cellLabelCol: {
+			type: Array as PropType<number[] | string[]>,
+			default() {
+				return [];
+			}
+		},
+		fillColorFn: {
+			type: Function,
+			default() {
+				return '#000000';
+			}
+		},
+		barColorFn: {
+			type: Function,
+			default() {
+				return '#000000';
+			}
+		},
+		lineColorFn: {
+			type: Function,
+			default() {
+				return '#000000';
+			}
+		},
+		backgroundColor: {
+			type: String,
+			default: '#FFFFFF'
+		},
+		cellMarginRow: {
+			type: [String, Array],
+			default: '2px'
+		},
+		cellMarginCol: {
+			type: [String, Array],
+			default: '2px'
+		}
+	},
+
+	// ---------------------------------------------------------------------------- //
+	// data                                                                         //
+	// ---------------------------------------------------------------------------- //
+
+	// ///////////////////////////////////////////////////////////////////////////////
+
+	data() {
+		return {
+			dataCellList: [] as object[],
+			dataRowList: [] as object[][], // e.g. [[row1col1Obj, row1col2Obj]]
+			dataColList: [] as object[][], // e.g. [[row1col1Obj, row2col1Obj]]
+			dataParameters: new Set() as Set<string>, // e.g. ["age", "height", "weight"]
+			dataParametersMin: {}, // e.g. {param1: 0, param2: 3}
+			dataParametersMax: {}, // e.g. {param1: 10, param2: 17}
+			labelRowList: [] as number[] | string[],
+			labelColList: [] as number[] | string[],
+			numRows: 0,
+			numCols: 0,
+
+			microRowArray: new Uint32Array(0) as Uint32Array,
+			microColArray: new Uint32Array(0) as Uint32Array,
+
+			uniforms: {} as Uniforms,
+			worldWidth: 0,
+			worldHeight: 0,
+
+			screenHeight: 0,
+			screenWidth: 0,
+
+			selectedCellList: [] as SelectedCell[], // e.g. [[startrow, startcol, endrow, endcol]]
+			selectedRows: [] as CellStatus[], // e.g. [0, 0, 1, 0]
+			selectedCols: [] as CellStatus[], // e.g. [0, 0, 1, 0]
+			selectedCell: Array(4) as SelectedCell, // e.g. [startrow, startcol, endrow, endcol]
+
+			update: 0,
+			move: 0,
+			enableDrag: false,
+			resizeObserver: null as unknown as ResizeObserver
+		};
+	},
+
+	// ---------------------------------------------------------------------------- //
+	// computed                                                                     //
+	// ---------------------------------------------------------------------------- //
+
+	// ///////////////////////////////////////////////////////////////////////////////
+
+	computed: {
+		screenAspectRatio(): number {
+			return this.screenHeight / this.screenWidth;
+		},
+
+		dataParametersArray(): string[] {
+			return [...this.dataParameters];
+		},
+
+		numSelectedRows(): number {
+			return this.selectedRows.reduce(
+				(acc: number, val: CellStatus) => acc + Number(val !== CellStatus.NONE),
+				0
+			);
+		},
+
+		numSelectedCols(): number {
+			return this.selectedCols.reduce(
+				(acc: number, val: CellStatus) => acc + Number(val !== CellStatus.NONE),
+				0
+			);
+		},
```

Comment:
I think it is more performant, and maybe more readable if you don't use accumulator and the hardcast to Number...maybe something like.

```
return this.selectedRows.filter(d => d !== CellStatus.NONE).length
```

Didn't try on browser, but the reduce is slower running in node

```
const list = [];
for (let i = 0; i< 300; i++) {
  if (Math.random() < 0.5) {
    list.push(0);
  } else {
    list.push(1);
  }
}


function getOnes() {
  return list.reduce((acc, val) => acc + Number(val !== 0), 0)
}

function getOnes_2() {
  return list.filter(d => d !== 0).length;
}


let start = 0;
let end = 0;

start = performance.now();
for (let i = 0; i < 100; i++) {
  getOnes();
}
end = performance.now();
console.log(end - start);

start = performance.now();
for (let i = 0; i < 100; i++) {
  getOnes_2();
}
end = performance.now();
console.log(end - start);
```

Reply:
hmm i just pasted your benchmark into the browser console a few times and the accumulator is slightly (0.2ms) faster the majority of the time. that performance difference is negligable tho (runtime roughly 1ms for both) and i do like the readability of the filter better. we will use that and change it back if it ever becomes an issue (probably not).

Reply:
Ah right, probably because it creates a new array - I have comparable result as yours when I increase the number of iterations.